library/core/src/any.rs

```diff
@@ -700,3 +770,284 @@ pub const fn type_name<T: ?Sized>() -> &'static str {
 pub const fn type_name_of_val<T: ?Sized>(_val: &T) -> &'static str {
     type_name::<T>()
 }
+
+///////////////////////////////////////////////////////////////////////////////
+// Provider trait
+///////////////////////////////////////////////////////////////////////////////
+
+/// Trait implemented by a type which can dynamically provide values based on type.
+#[unstable(feature = "provide_any", issue = "96024")]
+pub trait Provider {
+    /// Data providers should implement this method to provide *all* values they are able to
+    /// provide by using `demand`.
```

Comment:
suggestion: document the expected behaviour if you provide a value for the same type multiple times.  (Or explicitly say that's not allowed -- at which point you might consider panicking for that in `provide` and `provide_with`.)